api/trace/trace_context_propagator.go

```diff
@@ -29,21 +28,33 @@ const (
 	supportedVersion  = 0
 	maxVersion        = 254
 	traceparentHeader = "traceparent"
+	tracestateHeader  = "tracestate"
+)
+
+type traceContextPropagatorKeyType uint
+
+const (
+	tracestateKey traceContextPropagatorKeyType = 0
 )
 
 // TraceContext propagates SpanContext in W3C TraceContext format.
 //nolint:golint
 type TraceContext struct{}
 
 var _ propagation.HTTPPropagator = TraceContext{}
-var traceCtxRegExp = regexp.MustCompile("^[0-9a-f]{2}-[a-f0-9]{32}-[a-f0-9]{16}-[a-f0-9]{2}-?")
+var traceCtxRegExp = regexp.MustCompile("^([0-9a-f]{2})-([a-f0-9]{32})-([a-f0-9]{16})-([a-f0-9]{2})(?:-.*)?$")
 
 // DefaultHTTPPropagator returns the default trace HTTP propagator.
 func DefaultHTTPPropagator() propagation.HTTPPropagator {
 	return TraceContext{}
 }
 
 func (TraceContext) Inject(ctx context.Context, supplier propagation.HTTPSupplier) {
+	tracestate := ctx.Value(tracestateKey)
+	if tracestate != nil {
+		supplier.Set(tracestateHeader, tracestate.(string))
```

Comment:
this is always guaranteed to be a string, right, and thus can't panic a process?

Reply:
Since the `tracestateKey` is not exported, nor is the `traceContextPropagatorKeyType` type, and the `propagation.HTTPSupplier#Get` method returns a string, I'm confident that the only thing that could put a value there is going to put a string.

Reply:
As I was updating the regexp anyways, I added validation of the string coercion.